### PR TITLE
Add crumb exception for Oauth (applink) requests

### PR DESCRIPTION
**Pull request by dkjellin**
as they should be unprotected and have proper handling for these things already

### CODE REVIEW

**Review by alibk0rd — APPROVED** [review]

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/OauthCrumbExclusion.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * Adds exception to the CSRF protection filter for OAuth (applink) requests.
+ */
+@Extension
+public class OauthCrumbExclusion extends CrumbExclusion {
+
+    @Override
+    public boolean process(HttpServletRequest req, HttpServletResponse resp, FilterChain chain)
+            throws IOException, ServletException {
+        String pathInfo = req.getPathInfo();
+        if (isEmpty(pathInfo) || !OAuthRequestUtils.isOAuthAccessAttempt(req)) {
```
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/OauthCrumbExclusion.java:25]: Is the `isEmpty` required? `isOAuthAccessAttempt` is pulling out the path info, so I think the call is redundant.

Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/OauthCrumbExclusion.java:25]: Request tokens aren't part of this solution- do we provide credentials when hitting the request token endpoint?

Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/OauthCrumbExclusion.java:25]: I shamelessly stole the webhook one, and you're right, the `isEmpty` and the whole `pathInfo` is not needed, I'll update

Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/OauthCrumbExclusion.java:25]: As for request tokens, it is my understanding that they do not require permissions. @alibk0rd or @goravjoshi can confirm if my understanding is corerct

Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/OauthCrumbExclusion.java]: Also, I should point out, this is not about _credentials_ it is about disabling the XSRF protection for the relevant calls. Any authentication etc still needs to be handled in the relevant resources.

Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/OauthCrumbExclusion.java]: So just following up mostly for clarification (as have verified this works)- this is adding an exception for all requests that match our access token endpoint. The access token methods are served through the same action as our request tokens (`TokenEndpoint`) and I was wondering why request tokens seem to be working but access tokens require an exception? I assume the TokenEndpoint action, which is not unprotected, is where the 403 is coming from.

Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/OauthCrumbExclusion.java]: Not quite, this is adding an exception for the csrf protection (which is unrelated to any form of authentication/authorisation).  to requests that are deemed to be Oauth by way of checking the `isOauthAccessAttempt` return value. If it returns 'true' (it is an oAuth access attempt) the request is allowed through even if there is no CSRF headers present. It is then up to the receiving resource to do normal authentication/authorisation checks (which may of course be delegated to Jenkins via standard means). 

All this does is to say that 'if it is an oAuthAccessAttempt do not require the CSRF headers on the request' it does not do anything else, or disable anything else.

**Review by mhenschke-atl — APPROVED** [review]

**Pull request merged**